Assets/MixedRealityToolkit.Services/InputSystem/FocusProvider.cs

```diff
@@ -412,6 +412,97 @@ public override int GetHashCode()
             }
         }
 
+        /// <summary>
+        /// Helper class for managing the visibility of the gaze pointer to match windows mixed reality and HoloLens 2
+        /// When application starts, gaze pointer is visible. Then when articulate hands / motion controllers
+        /// appear, hide the gaze cursor. Whenever user says the voice wake word, make the 
+        /// gaze controller appear.
+        /// </summary>
+        private class GazePointerStateMachine : IMixedRealitySpeechHandler
+        {
+            private enum GazePointerState
+            {
+                Initial, // When the application starts up, the gaze pointer should be active
+                GazePointerActive, // Gaze pointer is active when no hands are visible, after "select" 
+                GazePointerInactive // Gaze pointer is inactive as soon as motion controller or articulated hand pointers appear
+            }
+            private GazePointerState gazePointerState = GazePointerState.Initial;
+            private bool activateGazeKeywordIsSet = false;
+
+            public bool IsGazePointerActive
+            {
+                get { return gazePointerState != GazePointerState.GazePointerInactive; }
+            }
+
+            public void UpdateState(HashSet<PointerData> pointers, GenericPointer currentGazePointer)
+            {
+                int numFarPointers = 0;
+                int numNearPointersActive = 0;
+                foreach (var p in pointers)
+                {
+                    if (p.Pointer != null)
```

Comment:
It's not enough to do a null check on a pointer.
This is because the pointer is stored as an interface, but many of them are actually monobehaviors.

When a pointer is destroyed, it will pass a regular null check, but throw a MissingReferenceException. You must also cast it to monobehavior to null check it properly. This unfortunately is much more expensive, but safer.

This is why I'm so adamant pointers should never be destroyed, so that we don't ever hold a list of null pointers.
https://github.com/microsoft/MixedRealityToolkit-Unity/issues/4223

Reply:
Hm, interesting. Since this does not regress prior behavior (we did the same null check previously), I think we could keep this in for now, and then fix it and other null checks later, as part of a separate issue. Trying to keep this scoped to implementing the specific behavior to match the shell.

Reply:
I understand wanting to keep the same behavior.

My experience with pointers however has shown that if the pointer is destroyed, it won't be considered null until unity's native side cleans it up.

This is why you want to know if the reference is still valid, which can only be done with the overridden monobebehavior null check, which isn't called unless you cast the pointer reference to monobebehavior.

Reply:
This is the same reason the `?` fails on monobebehaviors. This check is actually somewhat new since it wasn't possible before dot net 4.x. It works the same way as the ternary operator or the ?? Null check however.

You can read about how unity handles null checks here

https://blogs.unity3d.com/2014/05/16/custom-operator-should-we-keep-it/

Reply:
Removed this null check, because the pointer should never be null at this point in time. Other code in update loop already assumes the pointer is not null. It's appropriate in this case to crash/throw exception because this reflects program being in incorrect state.